### PR TITLE
Feat/bewust bezorgd

### PR DESCRIPTION
**Pull request by jdsmith022**
Implementation of Bewust Bezorgd, 
Use of plugin without comp button, and 
Only load middleware on WC pages

Bewust Bezorgd implemented for WC -> added function updateMerchant to sent GraphQL query to backend to request and return store's BB Api id and password which is then added as option

added check that page is either a WC page, cart or checkout page. If true, middleware is loaded

added switch for hidden feature of hiding comp button in store

### CODE REVIEW

File: co2ok-plugin.php [hunk 1]
```diff
@@ -243,6 +243,51 @@ final public static function remoteLogging($message = "Unspecified message.")
...
+        Co2ok_Plugin::remoteLogging(json_encode(["Logging inside updateMerchant to test scheduler"]));
+        $graphQLClient = new \co2ok_plugin_woocommerce\Components\Co2ok_GraphQLClient(Co2ok_Plugin::$co2okApiUrl);
+
+        $merchantId = get_option('co2ok_id');
+        $merchantSecret = get_option('co2ok_secret');
+
+        $graphQLClient->query(function ($query) use ($merchantId, $merchantSecret) {
+            $query->setFunctionName('merchant');
+            $query->setFunctionParams(array('id' => $merchantId, 'secret' => $merchantSecret));
+            $query->setFunctionReturnTypes(array("bbShopid", "bbPassword"));
```
Comment (Mil0dV) [inline co2ok-plugin.php:262]: The backend expects snake_case, then the query should get results

Comment (jdsmith022) [reply co2ok-plugin.php:262]: Did a check with this id and secret and received hoi back: 
Screenshot 2021-04-06 at 2 38 42 PM

File: Components/Co2ok-BewustBezorgd.php [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+			}
+		}
+
+		//check shipping method selected is accepted by BB api
+		$shippingChoice = $this->correctShipping($this->shippingMethod);
+
+		// Q&D PC cleanup
+		if (!preg_match("/[a-zA-Z]+$/", substr($this->shopPostCode, -2)))
+			$this->shopPostCode .= 'JS';
+
```
Comment (mvhou) [inline Components/Co2ok-BewustBezorgd.php:79]: I'm not 100% sure what's happening here. It checks if the last two characters in the shop's post code are letters, and if not it concatenates JS?

Comment (jdsmith022) [reply Components/Co2ok-BewustBezorgd.php:79]: Yes, there are sometimes issues with the postcode. BB API only works with NL post codes so we have a check for that, but somehow there a times when the postcode won't have letter to it...how the order gets through, we don't know, so this checks if there are letters attached to the post code, if not we attach JS.

File: co2ok-plugin.php [hunk 3]
```diff
@@ -595,6 +652,46 @@ final public function co2ok_load_plugin_textdomain()
         load_plugin_textdomain( 'co2ok-for-woocommerce', FALSE, basename( dirname( __FILE__ ) ) . '/languages/' );
     }
 
+
+    final private function co2ok_bewust_bezorgd($order, $merchantId) {
```
Comment (Mil0dV) [inline co2ok-plugin.php:656]: `co2ok_bewust_bezorgd()` doesn't say what this function _does_.

File: co2ok-plugin.php [hunk 4]
```diff
@@ -595,6 +652,46 @@ final public function co2ok_load_plugin_textdomain()
...
+            } else {
+                $orderDetails = array (
+                    'storePostCode' => $shopPostCode,
+                    'storeCountry' => $shopCountry,
+                    'destPostCode' => $destPostCode,
+                    'destCountry' => $destCountry,
+                    'weight' => $weight,
+                    'shippingMethod' => $shippingMethod,
+                    'merchantId' => $merchantId);
+                    Co2ok_Plugin::remoteLogging(json_encode(["Order delivery or store outside of NL. Delivery country ", $orderDetails]));
```
Comment (Mil0dV) [inline co2ok-plugin.php:685]: On second thought I don't think it's very useful to log this remotely. Let's disable #685 for now.

File: co2ok-plugin.php [hunk 5]
```diff
@@ -1088,13 +1193,13 @@ final public function co2ok_widget_shortcode($atts = []){
...
-        
-        $widget_code = 
+
+        $widget_code =
         '<div id="widgetContainerXL" style="width:auto;height:auto;display:flex;flex-direction:row;align-items:center;margin-top: 5px;"></div>'.
         // '<script src="http://localhost:8080/widget/co2okWidgetXL.js" ' .
         '<script src="https://co2ok.eco/widget/co2okWidgetXL-' . $code . '.js" ' .
-        'async div="widgetContainerXL" merchantId=' . $code . ' widgetSize="' . $size . '" widgetColor="' . $color . '" lang="' . $lang . '"></script>';
-        
+        'defer div="widgetContainerXL" merchantId=' . $code . ' widgetSize="' . $size . '" widgetColor="' . $color . '" lang="' . $lang . '"></script>';
```
Comment (Mil0dV) [inline co2ok-plugin.php:1201]: Ah, nice! I think we should also do #1179

**Review by Mil0dV — CHANGES_REQUESTED** [review]

File: Components/Co2ok-BewustBezorgd.php [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+			'MediumTimeframe',
+			'EveningDelivery',
+			'SameDay',
+			'SundayDelivery'
+		);
+		foreach ($shippingCategory as $method) {
+			if ($method == $shipping)
+				return $shipping;
+		}
+		return 'NextDay';
```
[the remark this answers is not in the document]
Comment (jdsmith022) [reply Components/Co2ok-BewustBezorgd.php:190]: added remote log for not match but will look futher into this detail on next PR

File: Components/Co2ok-BewustBezorgd.php [hunk 7]
```diff
@@ -103,7 +103,7 @@ public function store_order_to_bb_api($orderId) {
 				$emissionsGrams = $responseTwoLegs['emission'];
 				$diesel = $responseTwoLegs['metersDiesel'];
 				$gasoline = $responseTwoLegs['metersGasoline'];
-				Co2ok_BewustBezorgd::remoteLogging(json_encode(["Logging BB Api two-legs error response: ", $responseTwoLegs['errors'], "Emissions(g): ", $emissionsGrams, "Diesel: ", $diesel, "Gasoline: ", $gasoline]));
+				\co2ok_plugin_woocommerce\Co2ok_Plugin::remoteLogging("Logging BB Api two-legs error response: " . $responseTwoLegs['errors'], " Emissions(g): " . $emissionsGrams, "Diesel: " . $diesel . " Gasoline: ".  $gasoline);
```
Comment (Mil0dV) [inline Components/Co2ok-BewustBezorgd.php:106]: Don't we need json_encode anymore? We seem to be doing it everywhere in the code

File: co2ok-plugin.php (hunk 3 above)
Comment (Mil0dV) [reply co2ok-plugin.php:658]: I was commenting on the method name, not the lack of documentation ;)

**Review by Mil0dV — CHANGES_REQUESTED** [review]

File: co2ok-plugin.php (hunk 3 above)
Comment (jdsmith022) [reply co2ok-plugin.php:658]: oh haha

File: Components/Co2ok-BewustBezorgd.php (hunk 7 above)
Comment (jdsmith022) [reply Components/Co2ok-BewustBezorgd.php:106]: that's a good point. I did get a logs in the log entries when I tested, but let me check again

**Pull request merged**